### PR TITLE
Implement search by defendant name and date of birth

### PR DESCRIPTION
**Pull request by jsugarman**
#### What
Enable search against adaptor API by defendant name and DoB

#### Ticket

[CACP-117](https://dsdmoj.atlassian.net/browse/CACP-117)
[CACP-169](https://dsdmoj.atlassian.net/browse/CACP-169)

#### Why
Defendant name search as specified in [miro board designs](https://miro.com/app/board/o9J_kv-Sqxg=/).

The HMCTS Common platform API and, consequently, the adaptor
enable search by defendant name but require both the full name
and the DoB of the defendant in order to return results.

#### How
Add DoB field to the defendant name search
page and supply this to the adaptor in search
query submission. In addition a name parser
has been added to enable the user entered name
to be parse the first name and last name so that
these too can be passed to the adaptor in the
expected param format.

--------

#### TODO (wip)

 - [X] add dob field to defendant name search
 - [X] add name parser to handle user name entry

### CODE REVIEW

File: app/models/search_filter.rb [hunk 1]
```diff
@@ -3,7 +3,7 @@
 class SearchFilter
   include ActiveModel::Model
 
-  validates :id, presence: true
+  validates :id, presence: { message: 'Choose a filter' }
```
Comment (naseberry) [inline app/models/search_filter.rb]: i18n??

File: app/views/searches/_form.html.haml [hunk 2]
```diff
@@ -2,4 +2,10 @@
   = f.govuk_error_summary
   = f.hidden_field(:filter, value: @filter)
   = f.govuk_text_field(:term, value: @term, label: { text: @label }, hint_text: @hint )
+
+  - if @filter.eql?('defendant')
+    = f.govuk_date_field :dob,
+      legend: { text: 'Defendant date of birth' },
+      hint_text: 'For example, 31 3 1980'
```
Comment (naseberry) [inline app/views/searches/_form.html.haml]: locales?

File: app/models/search_filter.rb (hunk 1 above)
Comment (jsugarman) [reply app/models/search_filter.rb]: yep, was going to get on to that

File: app/views/searches/_form.html.haml (hunk 2 above)
Comment (jsugarman) [reply app/views/searches/_form.html.haml]: 👍

File: spec/features/search/case_reference_spec.rb [hunk 3]
```diff
@@ -10,7 +10,7 @@
   scenario 'with results', stub_case_reference_results: true do
     visit '/'
 
-    choose 'By case reference'
+    choose 'Search by case reference'
```
Comment (naseberry) [inline spec/features/search/case_reference_spec.rb]: Maybe use key from locales, therefore no need to update spec when locales value changes

File: spec/lib/court_data_adaptor/name_parser_spec.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    context 'when one term' do
+      let(:term) { 'fred   ' }
+
+      it 'has only first name' do
+        is_expected.to eql(first: 'fred', last: nil, middle: nil)
+      end
+    end
+
+    context 'when two terms' do
+      let(:term) { 'fred   alberts   ' }
```
Comment (naseberry) [inline spec/lib/court_data_adaptor/name_parser_spec.rb]: Is there a specific reason why multiple spaces are being used?

File: spec/features/search/case_reference_spec.rb (hunk 3 above)
Comment (jsugarman) [reply spec/features/search/case_reference_spec.rb]: For me the test actually tests the translation is working too

File: spec/lib/court_data_adaptor/name_parser_spec.rb (hunk 4 above)
Comment (jsugarman) [reply spec/lib/court_data_adaptor/name_parser_spec.rb]: yep, testing whitespace removal

**Pull request merged**